File: Problems/Algorithms/363. Max Sum of Rectangle No Larger Than K/max_rectangle_sum.py
#!/usr/bin/env python3

import math
from typing import List
from sortedcontainers import SortedList


class Solution:
    def maxSumSubmatrix(self, matrix: List[List[int]], k: int) -> int:
        m, n = len(matrix), len(matrix[0])
        ans = -math.inf
        for top in range(m):
            arr = [0] * n
            for bottom in range(top, m):
                for i in range(n):
                    arr[i] += matrix[bottom][i]
                ans = max(ans, self.max_sum_subarray(arr, n, k))            
        return ans

    def max_sum_subarray(self, arr: List[int], n: int, k: int) -> int:
        sums = SortedList([0])
        right = 0
        ans = -math.inf
        for i in range(n):
            right += arr[i]
            left = self.ceiling(sums, right - k)
            if left is not None:
                ans = max(ans, right - left)
            sums.add(right)
        return ans

    def ceiling(self, sorted_list: SortedList, k: int) -> int:
        idx = sorted_list.bisect_left(k)
        if idx < len(sorted_list):
            return sorted_list[idx]
        return None

File: Problems/Algorithms/363. Max Sum of Rectangle No Larger Than K/MaxRectangleSum.java
class Solution {
    public int maxSumSubmatrix(int[][] matrix, int k) {
        int m = matrix.length;
        int n = matrix[0].length;
        
        int ans = Integer.MIN_VALUE;
        for (int top = 0; top < m; top++) {
            int[] arr = new int[n];
            for (int bottom = top; bottom < m; bottom++) {
                for (int i = 0; i < n; i++) {
                    arr[i] += matrix[bottom][i];
                }
                ans = Math.max(ans, maxSumSubarray(arr, n, k));
            }
        }

        return ans;
    }

    private int maxSumSubarray(int[] array, int n, int k) {
        TreeSet<Integer> bst = new TreeSet<>();
        bst.add(0);

        int max = Integer.MIN_VALUE;
        int sum = 0;
        for (int i = 0; i < n; i++) {
            sum += array[i];
            Integer lower = bst.ceiling(sum - k);
            if (lower != null) {
                max = Math.max(max, sum - lower);
            }
            bst.add(sum);
        }

        return max;
    }
}
